dca/methods_comparison.py

```diff
@@ -610,3 +610,265 @@ def fit_transform(self, X, n_components=None):
             n_components = self.n_components
         self.fit(X)
         return self.transform(X, n_components=n_components)
+
+
+class JPCA(object):
+    """ jPCA.
+
+    Parameters
+    ----------
+    n_components : even int (default=6)
+        Number of components to reduce X to.
+
+    mean_subtract: boolean (default=True)
+        Whether to subtract the cross-condition mean from each condition
+        before running jPCA.
+
+    Attributes
+    ----------
+    eigen_vecs_ : list
+        List of numpy eigenvectors from JPCA skew symmetric matrix sorted in
+        descending order by magnitude of eigenvalue.
+
+    eigen_vals_ : list
+        List of eigenvalues from JPCA skew symmetric matrix. The index
+        of each eigenvalue corresponds with the eigenvector in eigen_vecs_.
+
+    pca_ : sklearn.decomp.PCA object
+        PCA object used to transform X to X_red.
+    """
+    def __init__(self, n_components=6, mean_subtract=True):
+        if n_components // 2 != n_components / 2:
+            raise ValueError("n_components must be even int")
+        self.n_components_ = n_components
+        self.mean_subtract = mean_subtract
+        self.eigen_vecs_ = None
+        self.eigen_vals_ = None
+        self.pca_ = None
+
+    def fit(self, X):
+        """ Fit a jPCA model to X.
+
+        Parameters
+        ----------
+        X : ndarray (time, features) or (conditions, time, features)
+            Data to fit using jPCA model.
+
+        Returns
+        -------
+        X_red : ndarray (time*conditions, n_components)
+            Dimensionally reduced and preprocessed X. X_red is returned
+            as self.transform requires the data matrix in this form.
+        """
+        if self.n_components_ > X.shape[1]:
+            raise ValueError("n_components is greater than number of features in X.")
+
+        if self.mean_subtract:
+            X = self._mean_subtract(X)
+
+        # stack conditions if more than 1:
+        if len(X.shape) == 3:
```

Comment:
I'd also like to treat multiple conditions slightly differently. I think a reasonable procedure is

1. Remove the mean per condition
2. Do PCA across all conditions
3. Calculate the diffs per condition
4. Concatenate the diffs to do jPCA

Reply:
Added two methods to address this.